File: WindowsMobileClient/Windows Mobile Client/Menus/ReadOnlyContextMenu.vb
Imports System.ComponentModel
Imports activiser.Library.Forms

<ComponentModel.DesignTimeVisible(True)> _
Public Class ReadOnlyContextMenu : Inherits ContextMenuBase

    Public Sub New()
        MyBase.New(ContextMenuType.Readonly)
    End Sub

    Private _showCall As Boolean = True
    Public Property ShowCall() As Boolean
        Get
            Return _showCall
        End Get
        Set(ByVal value As Boolean)
            _showCall = value
        End Set
    End Property

    Protected Overrides Sub OnPopup(ByVal e As System.EventArgs)
        MyBase.OnPopup(e)

        If Not ShowCall Then Return ' not relevant if we don't have the make phone call menu item.

#If WINDOWSMOBILE Then
        If _showCall AndAlso HavePhone() Then
            If Not Me.MenuItems.Contains(CallMI) Then
                Me.MenuItems.Add(CallMI)
            End If
            CallMI.Enabled = False
        Else
            If Me.MenuItems.Contains(CallMI) Then
                Me.MenuItems.Remove(CallMI)
            End If
        End If
#End If

        Try
            Dim tb As TextBox = TryCast(Me.SourceControl, TextBox)
            If tb IsNot Nothing Then
                SelectAllMI.Enabled = True
#If WINDOWSMOBILE Then
                CallMI.Enabled = CanPhone(tb)
#End If
                Return
            End If

            Dim nb As NumberPicker = TryCast(Me.SourceControl, NumberPicker)
            If nb IsNot Nothing Then
                SelectAllMI.Enabled = True
#If WINDOWSMOBILE Then
                CallMI.Enabled = CanPhone(nb.Text)
#End If            
                Return
            End If

            Dim lb As Label = TryCast(Me.SourceControl, Label)
            If lb IsNot Nothing Then
                SelectAllMI.Enabled = False
#If WINDOWSMOBILE Then
                CallMI.Enabled = CanPhone(lb.Text)
#End If            
                Return
            End If

            Dim cb As ComboBox = TryCast(Me.SourceControl, ComboBox)
            If cb IsNot Nothing Then
                SelectAllMI.Enabled = False
#If WINDOWSMOBILE Then
                If cb.SelectedText.Length <> 0 Then
                    CallMI.Enabled = CanPhone(cb.SelectedText)
                Else
                    CallMI.Enabled = CanPhone(cb.Text)
                End If
#End If            
                Return
            End If

            SelectAllMI.Enabled = False
#If WINDOWSMOBILE Then
            CallMI.Enabled = False
#End If            

        Catch 'ex As Exception
#If WINDOWSMOBILE Then
            CallMI.Enabled = False
#End If            
        End Try
    End Sub
End Class

File: Library/activiser.Library.Forms/ReadOnlyContextMenu.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;
using System.Windows.Forms;

namespace activiser.Library.Forms
{
    [DesignTimeVisible(true),DesignerCategory("Component")]
    public class ReadOnlyContextMenu : ContextMenu
    {
        private System.Windows.Forms.MenuItem CopyMI;
        private System.Windows.Forms.MenuItem SelectAllMI;
        private System.Windows.Forms.MenuItem CallMI;

        internal ReadOnlyContextMenu()
            : base()
        {
            CopyMI = new System.Windows.Forms.MenuItem();
            CopyMI.Text = Properties.Resources.ContextMenuCopy;
            this.MenuItems.Add(CopyMI);

            SelectAllMI = new System.Windows.Forms.MenuItem();
            SelectAllMI.Text = Properties.Resources.ContextMenuSelectAll;
            this.MenuItems.Add(SelectAllMI);

            // always instantiate, just only add when needed.
            CallMI = new System.Windows.Forms.MenuItem();
            CallMI.Text = Properties.Resources.ContextMenuCall;
        }

        private bool _showCall = false;
        public bool ShowCall
        {
            get { return _showCall; }
            set
            {
                _showCall = value;
                if (_showCall && !this.MenuItems.Contains(CallMI))
                    this.MenuItems.Add(CallMI);
                else if (this.MenuItems.Contains(CallMI)) 
                    this.MenuItems.Remove(CallMI);
            }
        }

        protected override void OnPopup(System.EventArgs e)
        {
            base.OnPopup(e);

            
            //bool enablePhone = false;
            //bool enableSelectAll = false;
            ////Dim enableMail As Boolean 

            //if (!TestTextBox(this.SourceControl as TextBox, ref enablePhone, ref enableSelectAll))
            //    if (!TestNumberPicker(this.SourceControl as NumberPicker, ref enablePhone, ref enableSelectAll))
            //        if (!TestComboBox(this.SourceControl as ComboBox, ref enablePhone, ref enableSelectAll))
            //            if (!TestDateTimePicker(this.SourceControl as DateTimePicker, ref enablePhone, ref enableSelectAll))
            //                TestLabel(this.SourceControl as Label, ref enablePhone, ref enableSelectAll);

            //ContextMenuBase.SelectAllMI.Enabled = enableSelectAll;

            //if (ShowCall) ContextMenuBase.CallMI.Enabled = enablePhone;
        }

        private static bool TestTextBox(TextBox tb, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (tb == null) return false;
            enableSelectAll = true;
            enablePhone = Phone.CanPhone(tb);
            return true;
        }

        private static bool TestNumberPicker(NumberPicker nb, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (nb == null)
                return false;
            enableSelectAll = true;
            if (string.IsNullOrEmpty(nb.SelectedText))
                enablePhone = Phone.CanPhone(nb.SelectedText);
            else
                enablePhone = Phone.CanPhone(nb.ToString());

            return true;
        }

        private static bool TestComboBox(ComboBox cb, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (cb == null)
                return false;
            enableSelectAll = false;
            enablePhone = Phone.CanPhone(cb.Text);
            return true;
        }

        private static bool TestDateTimePicker(DateTimePicker dt, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (dt == null)
                return false;
            enablePhone = false;
            enableSelectAll = false;
            return true;
        }

        private static bool TestLabel(Label lb, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (lb == null)
                return false;
            enableSelectAll = false;
            enablePhone = Phone.CanPhone(lb.Text);
            return true;
        }
    } 
}
